lib/cn/csched_sp3.c

```diff
@@ -947,13 +949,20 @@ sp3_dirty_node_locked(struct sp3 *sp, struct cn_tree_node *tn)
              * node has 3% garbage.
              * We use alen as the secondary discriminant to prefer nodes
              * with higher total bytes of garbage.
-             *
-             * TODO: The garbage caculation needs help:  It sometimes returns
-             * a non-zero result for nodes consisting entirely of unique keys.
              */
-            if (garbage > 1 && nkvsets > 1 && keys > keys_uniq) {
+            if (garbage > 0) {
                 weight = ((uint64_t)garbage << 32) | (cn_ns_alen(&tn->tn_ns) >> 20);
 
+                /* Accelerate GC if the preponderance of keys are tombs.
+                 * Use 97% to account for observed hlog estimation error,
+                 * but would hlog-provided error bounds be better?
+                 */
+                if (cn_ns_tombs(&tn->tn_ns) * 100 > keys_uniq * 97) {
+                    sp3_node_remove(sp, spn, wtype_length);
+                    weight *= 2;
+                    ev_debug(1);
```

Comment:
If I read this right, it amounts to:  number of tombs is about equal to number of keys w/o tombs.  Why not just compare number of tombs and number of non-tomb keys directly and not even worry about hyperloglog?

Reply:
I originally tried (tombs * 2 >= keys) but I could have a ton of tombs that do not annihilate any keys, or a ton of tombs that all kill just one key, etc,,  I'd rather not kick of a GC just to find that out...  But I'm open to suggestions, perhaps I'm not following you...

Reply:
Yeah... no easy way to know if tombs actually intersect with keys.

Reply:
Drat, was hoping you had some magic...

Reply:
It would be interesting for compaction to add stats to output kvsets on the output-key-count/input-tomb-count and output-key-count/input-ptomb-count (or something similar) to see if those stats are reasonably predictive for the next compaction in the node.

Reply:
Hmm... I think that should be:

- (inputKeyCount - outputKeyCount) / inputTombCount which is keys annihilated per tomb, and
- (inputKeyCount - outputKeyCount) / inputPTombCount which is keys annihilated per ptomb

Of course that's a problem in the case where, for example, I have only ptomb in the first compaction and only tomb in the following.  And it's not clear which to use when you have both in the first and following compaction.

In any case, the point is we might be able to improve annihilation estimates based on past annihilation history for a given node.  We'd have to prototype it and run a bunch of workloads to see if this approach has any merit.